File: go/102BinaryTreeLevelOrderTraversal.go
package main

import (
	//"container/list"
	"fmt"
)

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func levelOrder(root *TreeNode) [][]int {
	output := [][]int{}
	pOutput := &output

	var calculateChildrenLayer func(int, *TreeNode)
	calculateChildrenLayer = func(parentLayer int, curNode *TreeNode) {
		if curNode == nil {
			return
		}

		curLayer := parentLayer + 1
		if curLayer < len(*pOutput) {
			(*pOutput)[curLayer] = append((*pOutput)[curLayer], curNode.Val)
		} else if curLayer == len(*pOutput) {
			*pOutput = append(*pOutput, []int{curNode.Val})
		}

		if curNode.Left != nil {
			calculateChildrenLayer(curLayer, curNode.Left)
		}
		if curNode.Right != nil {
			calculateChildrenLayer(curLayer, curNode.Right)
		}
	}

	calculateChildrenLayer(-1, root)

	return output
}

func main() {
	output := levelOrder(
		&TreeNode{
			Val: 3,
			Left: &TreeNode{
				Val:   9,
				Left:  nil,
				Right: nil,
			},
			Right: &TreeNode{
				Val: 20,
				Left: &TreeNode{
					Val:   15,
					Left:  nil,
					Right: nil,
				},
				Right: &TreeNode{
					Val:   7,
					Left:  nil,
					Right: nil,
				},
			},
		})
	output1 := levelOrder(nil)
	output2 := levelOrder(
		&TreeNode{
			Val:   1,
			Left:  nil,
			Right: nil,
		})
	output3 := levelOrder(
		&TreeNode{
			Val: 1,
			Left: &TreeNode{
				Val:   2,
				Left:  nil,
				Right: nil,
			},
			Right: nil,
		})
	output4 := levelOrder(
		&TreeNode{
			Val:  1,
			Left: nil,
			Right: &TreeNode{
				Val:   2,
				Left:  nil,
				Right: nil,
			},
		})
	output5 := levelOrder(
		&TreeNode{
			Val: 3,
			Left: &TreeNode{
				Val:   1,
				Left:  nil,
				Right: nil,
			},
			Right: &TreeNode{
				Val:   2,
				Left:  nil,
				Right: nil,
			},
		})

	fmt.Println("\n******** output *********\n", output, output1, output2, output3, output4, output5)
}

File: cpp/102BinaryTreeLevelOrderTraversal.cpp
#include <iostream>
#include <set>
#include <vector>
#include <iterator>
#include <utility>
#include <cstdlib>
#include <limits>
#include <functional>
#include <stack>
#include <unordered_set>
#include <queue>

using namespace std;

template<class Container>
void printContainer(const Container& container) {
//    return;

    if (container.size() == 0) {
        cout << "[]" << endl;
        return;
    }

    for (auto it = container.begin(); it != container.end(); it++) {
        if (it == container.begin()) {
            cout << "[";
        }
        cout << *it;
        if (it == prev(container.end())) {
            cout << "]" << endl;
        } else {
            cout << ", ";
        }
    }
}

template<class T>
void print2DimVec(const vector<vector<T>>& container) {
//    return;

    if (container.size() == 0) {
        cout << "[]" << endl;
        return;
    }

    for (auto it = container.begin(); it != container.end(); it++) {
        if (it == container.begin()) {
            cout << "[" << endl;
        }
        printContainer(*it);
        if (it == prev(container.end())) {
            cout << "]" << endl;
        }
    }
}

/**
 * Definition for a binary tree node.
 */
 struct TreeNode {
     int val;
     TreeNode *left;
     TreeNode *right;
     TreeNode() : val(0), left(nullptr), right(nullptr) {}
     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 };

class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        vector<vector<int>> output;

        if (root == nullptr) {
            return output;
        }

        function<void(int, TreeNode*)> levelTraverse =
            [&output, &levelTraverse](int parentHeight, TreeNode* node) {
                if (node == nullptr) {
                    return;
                }

                int height = parentHeight + 1;
                if (height < output.size()) {
                    output[height].push_back(node->val);
                } else if (height == output.size()) {
                    output.push_back(vector<int>{node->val});
                }

                if (node->left != nullptr) {
                    levelTraverse(height, node->left);
                }
                if (node->right != nullptr) {
                    levelTraverse(height, node->right);
                }
        };

        levelTraverse(-1, root);

        return output;
    }
};


int main(){
    Solution s;
    TreeNode t7(7);
    TreeNode t15(15);
    TreeNode t20(20, &t15, &t7);
    TreeNode t9(9);
    TreeNode root(3, &t9, &t20);
    auto output = s.levelOrder(&root);

    auto output1 = s.levelOrder(nullptr);

    TreeNode root2(1);
    auto output2 = s.levelOrder(&root2);

    TreeNode t3(2);
    TreeNode root3(1, &t3, nullptr);
    auto output3 = s.levelOrder(&root3);


    TreeNode t4(2);
    TreeNode root4(1, nullptr, &t4);
    auto output4 = s.levelOrder(&root4);

    TreeNode t5(1);
    TreeNode t6(2);
    TreeNode root5(3, &t5, &t6);
    auto output5 = s.levelOrder(&root5);

    cout << "*************output*************" << endl;
    print2DimVec(output);
    print2DimVec(output1);
    print2DimVec(output2);
    print2DimVec(output3);
    print2DimVec(output4);
    print2DimVec(output5);

    return 0;
}
